generate dev environment, add <vue-vega> component, and add example

aantn opened the pull request:
A few comments:

1. Run `npm run serve` to see a working application that uses this library. It demoes the very basic <vue-vega> component that I added on top of the existing composition API.
2. This is a work in progress. I'm opening this PR to get early feedback.
3. Dev tooling: to be honest, I have very little experience here. I generated some tooling automatically using https://www.npmjs.com/package/vue-sfc-rollup but feel free to throw it all away and replace it with something else. I would rather not deep dive into Javascript tooling right now, if it's avoidable.

TODOs:

- [ ] I'm temporarily converting various types to `any` to fix typing errors in the existing code. This needs to be fixed.
- [ ] Verify that the rollup config works and includes not only components but also the composition API parts. Test importing this library in another app.
- [ ] Either use or remove VueVegaConfig.data. (Right now data is taken from the VueVegaConfig.spec.)

aantn committed:
generate dev environment, add <vue-vega> component, and add example

dev tooling was automatically generated using https://www.npmjs.com/package/vue-sfc-rollup

domoritz commented:
@jsbroks please review

jsbroks commented on build/rollup.config.js:
```diff
@@ -0,0 +1,180 @@
+import PostCSS from 'rollup-plugin-postcss';
+import { terser } from 'rollup-plugin-terser';
+import ttypescript from 'ttypescript';
+import typescript from 'rollup-plugin-typescript2';
```
Whats wrong with using [@rollup/plugin-typescript](https://www.npmjs.com/package/@rollup/plugin-typescript)

aantn replied:
Nothing. This was the default that https://www.npmjs.com/package/vue-sfc-rollup pulled in, but I've replaced it.

jsbroks commented on src/components/vue-vega.vue:
```diff
@@ -0,0 +1,35 @@
+      data: props.data,
+    };
+
+    const { render, loading } = useVega(config as any);
```
You should be able to set the correct typing on the props with [`PropType`](https://vuejs.org/v2/guide/typescript.html#Annotating-Props)

jsbroks commented on src/composition/useVega.ts:
```diff
@@ -33,7 +33,7 @@ export function useVega(config: VueVegaConfig) {
     config
   )
-  const { embed, ...vegaEmbed } = useVegaEmbed(el, spec, embedOptions)
+  const { embed, ...vegaEmbed } = useVegaEmbed(el, spec as any, embedOptions)
```
Why do we need the as any? We should reframe from using it when possible

jsbroks commented on src/composition/useVega.ts:
```diff
@@ -45,7 +45,8 @@ export function useVega(config: VueVegaConfig) {
       return output
     } catch (error) {
-      options?.onError?.(error)
+      console.log("error is ", error);
```
Probably best to remove console the log by default.

jsbroks commented on src/composition/useVega.ts:
```diff
@@ -45,7 +45,8 @@ export function useVega(config: VueVegaConfig) {
     } catch (error) {
-      options?.onError?.(error)
+      console.log("error is ", error);
+      options?.onError?.(error as any)
```
as `any`?

jsbroks commented on src/composition/useVega.ts:
```diff
@@ -55,7 +56,7 @@ export function useVega(config: VueVegaConfig) {
   // Watchers that update vega in the most optimal way by either modifying or
   // re-rendering
-  useVegaUpdater(spec, embedOptions, signals, data, render, modify)
+  useVegaUpdater(spec as any, embedOptions, signals, data, render, modify)
```
as any

jsbroks commented on src/composition/useVegaEmbed.ts:
```diff
@@ -20,7 +20,11 @@ export function useVegaEmbed(
 
   const embed = async () => {
-    if (el.value == null) return
+    console.log("embed called")
+    if (el.value == null) {
+      console.log("its null")
+      return
```
console logs

jsbroks reviewed (changes requested):
Nice job

domoritz commented:
Thank you for the review @jsbroks and thank you @aantn for the pull request.

aantn committed:
update list of packaged files

aantn committed:
fix build

aantn commented:
Thank you for the comments. Will fix and update the PR.

I have two rollup related questions I'm hoping someone can help with.

1. Something seems wrong with the rollup output in dist/ after running `npm run build`. File sizes are over 800kb. I understand that some of the files there are bundling up all dependencies, but shouldn't there be a bundle without dependencies that you can import in normal webpack applications? Again, I'm not familiar with Javascript packaging, so I apologize for my ignorance on how this should be setup.
2. I've had to workaround errors like `The inferred type of this node exceeds the maximum length the compiler will serialize. An explicit type annotation is needed.`. Is this a known issue due to the size of the vue-embed VisualizationSpec class? Is there an elegant workaround for this?

aantn committed:
fix build

1. fix ESM build
2. switch to standard rollup typescript plugin
3. remove broken non ESM builds

aantn committed:
change javascript targets to es6

aantn commented:
Alright, I've fixed the former issue with massive build sizes. Looks like the defaults from https://www.npmjs.com/package/vue-sfc-rollup weren't sane after all.

I've removed the non-ESM builds as they were broken.

aantn committed:
simplify rollup config

aantn committed:
generate types and update package.json metadata

aantn committed:
more work on fixing build - use rollup-plugin-ts just like vega-embed

aantn committed:
fix path to types

aantn committed:
simplify entrypoint and fix various issues

aantn commented:
I've gotten this closer to working but it is still somewhat broken. It looks like there is a fundamental issue using vue-demi for component libraries and there needs to be two versions of vue-vega - one for vue2 and one for vue3: https://github.com/vuejs/rollup-plugin-vue/issues/427

aantn commented:
I wont have time to work on this in the near future, but I did find an example project that is using Vue-Demi for a component library. If I or someone else picks this up, it's probably worth looking at: https://github.com/open-data-plan/g2plot-vue (see package.json and the scripts/postinstall.sh). This doesn't use rollup though.

domoritz commented:
@jsbroks @aantn could you help with updating the vega-vue library to the latest Vega and Vega-Lite? As for simplifications, we now use npm instead of yarn for package installation and also switched to esm only (which means not bundling the dependencies into the esm bundle but referencing them) which addresses a concern raised above. 

I'm not sure whether you want to do this with this pull request or close this pull request. Could you help, though, maybe by the end of July?
